8336492: Regression in lambda serialization

mcimadamore opened the pull request:
This PR consolidates the code for dealing with local captures in both `Lower` and `LambdaToMethod`. It does so by adding a new tree scanner, namely `CaptureScanner`, which is then subclassed by `Lower.FreeVarCollector` and also by the new `LambdaToMethod.LambdaCaptureScanner`.

The main idea behind the new visitor is that we can compute the set of (most) captured locals w/o relying on complex state from `Lower`, and make the logic general enough to work on *any* tree. This can be done by keeping track of all local variable declarations occurring in a given subtree `T`. Then, if `T` contains a reference to a local variable that has not been seen, we can assume that this variable is a captured variable. This simple logic works rather well, and doesn't rely on checking e.g. that the accessed variable has the same owner of the method that owns a local class (which then caused other artifacts such as `Lower::ownerToCopyFreeVarsFrom`, now removed).

The bigger rewrite is the one in `LambdaToMethod`. That class featured a custom visitor called `LambdaAnalyzerPreprocessor` (now removed) which maintains a stack of frames encountered during a visit, and tries to establish which references to local variables needs to be captured by the lambda, as well as whether `this` needs to be referenced by the lambda. Thanks to the new `CaptureScanner` visitor, all this logic can be achieved in a very small subclass of `CaptureScanner`, namely `LambdaCaptureScanner`.

This removes the need to keep track of frames, and other ancillary data structures. `LambdaTranslationContext` is now significantly simpler, and its most important field is a `Map<VarSymbol, VarSymbol>` which maps logical lambda symbols to translated ones (in the generated lambda method). We no longer need to maintain different maps for different kinds of captured symbols.

The code for patching identifiers in a lambda expression also became a lot simpler: we just check if we have pending lambda translation context and, if so, we ask if the identifier symbol is contained in the translation map. Otherwise, we leave the identifier as is.

Fixing the issue referenced by this PR is now very simple: inside `LambdaCaptureScanner`, we just need to make sure that any identifier referring to a captured local field generated by `Lower` is re-captured, and the rest just works. To make the code more robust I've added a new variable flag to denote synthetic captured fields generated by `Lower`.

#### Compatibility

This PR changes the set of fields that are added to local/anonymous classes. Consider the following case:

```
class Outer {
    void m(int x) {
        class Local extends Serializable {
            void print() {
                System.out.println(x);
            }
        }
    }
}
```

This used to generate the following bytecode for `Local`:

```
class Outer$1Local {
  final int val$x;
    descriptor: I
  final Outer this$0;
    descriptor: LOuter;
  Outer$1Local();
    descriptor: (LOuter;I)V

  void print();
    descriptor: ()V
}
```

With this patch, we only generate this:

```
class Outer$1Local implements java.io.Serializable {
  final int val$x;
    descriptor: I
  Outer$1Local();
    descriptor: (LOuter;I)V

  void print();
    descriptor: ()V
}
```

That is, the field for `this$0` is omitted (as that's not used). This affects the serialized form of this local class, as there's now one less field in the serialized form. While this change is desirable (it brings behavior in sync with lambda expressions, and avoiding capture of enclosing instance might be beneficial for GC reachability), we need to make sure this is ok.

Also, it is possible that the _order_ in which local capture fields appear in classes/constructor parameters of local/anon classes changes slightly, due the fact we're using a different visitor.

Another, minor issue in this rewrite is that now generated methods for non-serializable lambdas can have different names from before. In the old code, a shared counter for _all_ lambdas in a class was used. As the new code shares the same logic for checking name clashing as the one used in serializable lambdas, we now increment the suffix after `lambda$foo$` if needed. So if two lambdas are defined in two methods `foo` and `bar`, we now get `lambda$foo$0` and `lambda$foo$bar$0` (before, one of them would have had the index set to `1`). While this affects some tests, which name we generate for synthetic lambda methods is a private contract, so I didn't think that preserving 100% compatibility was relevant here.

@cushon perhaps you could help assessing the compatibility impact of this change (as you seem to have a codebase which rely on serialization of local classes/lambdas) ?

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Change requires CSR request [JDK-8337558](https://bugs.openjdk.org/browse/JDK-8337558) to be approved
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8336492](https://bugs.openjdk.org/browse/JDK-8336492): Regression in lambda serialization (**Bug** - P3)
 * [JDK-8337558](https://bugs.openjdk.org/browse/JDK-8337558): Regression in lambda serialization (**CSR**)


### Reviewers
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20349/head:pull/20349` \
`$ git checkout pull/20349`

Update a local copy of the PR: \
`$ git checkout pull/20349` \
`$ git pull https://git.openjdk.org/jdk.git pull/20349/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20349`

View PR using the GUI difftool: \
`$ git pr show -t 20349`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20349.diff">https://git.openjdk.org/jdk/pull/20349.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20349#issuecomment-2252774221)

mcimadamore commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Lower.java:
```diff
@@ -2337,7 +2243,7 @@ private boolean shouldEmitOuterThis(ClassSymbol sym) {
         return true;
       }
-      if (rs.isSerializable(sym.type)) {
+      if (sym.owner.kind != MTH && rs.isSerializable(sym.type)) {
```
Note, this change serialized form for local/anon classes, by omitting capture of enclosing `this` where possible. (see discussion on serialization compatibility in the PR).

mcimadamore replied:
Note: consistency with the lambda case is kind of nice, but if this turns out to be problematic, we can omit this change from this PR, and always capture `this$0` in case of a serializable local/anon class.

mcimadamore commented:
/csr needed

archiecobbs commented:
Refactoring Awesomeness Ratio (not counting unit tests) = 1.58. 👍

cushon commented:
This looks exciting!

> @cushon perhaps you could help assessing the compatibility impact of this change (as you seem to have a codebase which rely on serialization of local classes/lambdas) ?

I'd be happy to. I have started a round of testing at https://github.com/openjdk/jdk/pull/20349/commits/76a972b775504f06b1db183a1b3fd047c2507f85, it'll take about a day to have results.

One caveat is that Google's use of serialization generally doesn't rely on persisting the serialized form, and we had disabled the check for serialization in `shouldEmitOuterThis` so the enclosing instance could be skipped even for serializable classes. That has been fine for us, but I don't know how well that generalizes. There's some previous discussion about the compatibility impact of omitting `this` references in [JDK-8277965](https://bugs.openjdk.org/browse/JDK-8277965) and linked issues.

vicente-romero-oracle reviewed (commented):
great simplification, some comments for your consideration

vicente-romero-oracle commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/LambdaToMethod.java:
```diff
@@ -916,810 +1004,330 @@ private JCExpression makeIndyCall(DiagnosticPosition pos, Type site, Name bsmNam
+                    ret = new VarSymbol((sym.flags() & FINAL) | PARAMETER, sym.name, types.erasure(sym.type), translatedSym);
+                    ret.pos = sym.pos;
+                    // Set ret.data. Same as case LOCAL_VAR above.
+                    if (sym.isExceptionParameter()) {
```
do we really need to set this exception parameter thing for a method param?

mcimadamore replied:
I'm not 100% sure, but eyeballing the code, I see that `sym.isExceptionParameter` is used in the backend (e.g. `Code`) so I'd suspect that a failure in doing this would lead to issues when generating bytecode, which I suspect is why this code was added in the first place. But, while that seems useful for a local variable declared inside a lambda, I don't see how a lambda parameter can itself be an exception parameter...

vicente-romero-oracle commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/LambdaToMethod.java:
```diff
@@ -289,9 +336,18 @@ public JCTree translateTopLevelClass(Env<AttrContext> env, JCTree cdef, TreeMake
     @Override
     public void visitClassDef(JCClassDecl tree) {
         KlassInfo prevKlassInfo = kInfo;
```
future improvement? we have class info, lambda context, capture site info, should we probably gather them all in an environment like structure?

mcimadamore replied:
yes and no. First, note that capture site info is gone.
As for lambda context and klass info, we use them in different ways: the lambda context is set when we're in a lambda. The klass info is set when we are inside a klass. We create many lambda contexts for each klass info. We could of course move the lambda context field from the visitor to klass info, but I'm not sure how much that buys.

cushon commented:
Some initial testing shows a regression in type annotation handling with this change.

Given an example like the following, the type annotation in the lambda is now being incorrectly propagated to the constructor. A `RuntimeInvisibleTypeAnnotations` attribute for the annotation is emitted on both the synthetic lambda method and the constructor. I noticed this because the `length` for the `RuntimeInvisibleTypeAnnotations` is longer than the constructor, and both ASM and javap report an error for the class file.

```java
import java.lang.annotation.ElementType;
import java.lang.annotation.Target;
import java.util.ArrayList;
import java.util.List;

class T {
  @Target(ElementType.TYPE_USE)
  @interface A {}

  Runnable r =
      () -> {
        List<@A Object> xs = new ArrayList<>();
        xs.add(1);
        xs.add(2);
        xs.add(3);
        xs.add(5);
        xs.add(6);
        xs.add(7);
        xs.add(8);
      };

  T() {}
}
```

javap reports an error because the type annotation's length is longer than the constructor:

```
$ javap -c T
Compiled from "T.java"
class T {
  java.lang.Runnable r;

  T();
    Code:
Error: error at or after byte 0
Error: Fatal error: Bytecode offset out of range; bci=89, codeLength=14
```

Using the JDK 11 javap shows

```
  T();
    descriptor: ()V
    flags: (0x0000)
    Code:
      stack=2, locals=1, args_size=1
         0: aload_0
         1: invokespecial #1                  // Method java/lang/Object."<init>":()V
         4: aload_0
         5: invokedynamic #7,  0              // InvokeDynamic #0:run:()Ljava/lang/Runnable;
        10: putfield      #11                 // Field r:Ljava/lang/Runnable;
        13: return
      LineNumberTable:
        line 22: 0
        line 10: 4
        line 22: 13
      RuntimeInvisibleTypeAnnotations:
        0: #35(): LOCAL_VARIABLE, {start_pc=8, length=81, index=0}, location=[TYPE_ARGUMENT(0)]
          T$A
```

mcimadamore commented:
> Some initial testing shows a regression in type annotation handling with this change.

Thanks, I'll take a look

mcimadamore commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/LambdaToMethod.java:
```diff
@@ -338,22 +401,15 @@ public void visitLambda(JCLambda tree) {
                     sym::setTypeAttributes);
 
+            apportionTypeAnnotations(tree,
+                    owner::getInitTypeAttributes,
```
> not sure that this is semantically the same as before

Yeah, this is a bit tricky. Basically, the issue with type annotations is that they are encoded in different places depending on the case. E.g. a local variable inside a static/instance initializer has the type annotation recorded in the _class symbol_ (!!). There's also different methods to fetch type annotations, depending on whether it was a normal type annotation, or one in a static/instance initializer.

The general idea is that we try to figure out the symbol that "owns" the type annotation while visiting, and then move all lambda annotations from that owner onto the new lambda method symbol.

mcimadamore replied:
Should be better after [this push](https://github.com/openjdk/jdk/pull/20349/commits/29910e6c0bae0b75e386169e114a4849ed931436)

mcimadamore commented:
> > Some initial testing shows a regression in type annotation handling with this change.
> 
> Thanks, I'll take a look

Ugh. This is quite ugly. The type annotation is seen, and correctly moved by `LambdaToMethod`. Unfortunately, the annotation seems to be added in _two_ places: in the local variable, and then in the synthetic constructor of `T`. We move the former, but not the latter.

mcimadamore commented:
After trying different approaches, I realized that the cleanest (by far) solution is to simply capture the lambda owner symbol during `Attr`, instead of trying to reconstruct this owner using various (wrong) heuristics. When doing this, I realized that there were several issues which predated my refactoring - that is, the logic in `TypeAnnotations` relies on the `BLOCK` flag to be set on static/instance initializer owner method symbols - but sometimes this flag was missing (this is now fixed). Also, the `EnclosingMethodAttribute` sometimes reported `<clinit>` being the method owner of a local class. This is deliberately against the JVMS, which state:

>  In particular, method_index must be zero if the current class was immediately enclosed in source code by an instance initializer, static initializer, instance variable initializer, or class variable initializer. (The first two concern both local classes and anonymous classes, while the last two concern anonymous classes declared on the right hand side of a field assignment.) 

We even had a test which enforced the *wrong* behavior (!!).

Last, the static/instance initializer symbol used to have a `null` type. This is now fixed, as we can always rely on the `BLOCK` flag instead (which is, again, cleaner).

As a result, `CaptureSiteInfo` is now gone.

mcimadamore commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java:
```diff
@@ -1419,9 +1419,10 @@ public void visitBlock(JCBlock tree) {
             Symbol fakeOwner =
                 new MethodSymbol(tree.flags | BLOCK |
-                    env.info.scope.owner.flags() & STRICTFP, names.empty, null,
+                    env.info.scope.owner.flags() & STRICTFP, names.empty, fakeOwnerType,
```
the type was left null here, which caused some crashes in `LambdaToMethod`. There was really no reason for this to be null, except that `ClassWriter` looks at type being null as a test to see if owner is a static/instance init block. But we can use `BLOCK` for that (as `TypeAnnotations` does)

mcimadamore commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java:
```diff
@@ -3572,7 +3573,7 @@ public Env<AttrContext> lambdaEnv(JCLambda that, Env<AttrContext> env) {
                         Type clinitType = new MethodType(List.nil(),
                                 syms.voidType, List.nil(), syms.methodClass);
-                        clinit = new MethodSymbol(STATIC | SYNTHETIC | PRIVATE,
+                        clinit = new MethodSymbol(BLOCK | STATIC | SYNTHETIC | PRIVATE,
```
Added missing `BLOCK`, so that we're now consistent with what the rest of the code (not only `LambdaToMethod`) expects.

mcimadamore commented on src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/ClassWriter.java:
```diff
@@ -329,7 +329,7 @@ protected int writeEnclosingMethodAttribute(Name attributeName, ClassSymbol c) {
         ClassSymbol enclClass = c.owner.enclClass();
         MethodSymbol enclMethod =
-            (c.owner.type == null // local to init block
+            ((c.owner.flags() & BLOCK) != 0 // local to init block
```
new test, which looks at the `BLOCK` flag

mcimadamore commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/LambdaToMethod.java:
```diff
@@ -1731,17 +1339,15 @@ final class ReferenceTranslationContext extends TranslationContext<JCMemberRefer
-        CAPTURED_THIS;  // class symbols to translated synthetic parameters (for captured member access)
+        CAPTURED_VAR;   // variables in enclosing scope to translated synthetic parameters
 
         boolean propagateAnnotations() {
```
Addressed

mcimadamore commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/LambdaToMethod.java:
```diff
@@ -916,810 +1004,330 @@ private JCExpression makeIndyCall(DiagnosticPosition pos, Type site, Name bsmNam
+    /**
+     * Simple subclass modelling the translation context of a method reference.
+     */
+    final class ReferenceTranslationContext extends TranslationContext<JCMemberReference> {
```
Addressed [here](https://github.com/openjdk/jdk/pull/20349/commits/fcbbfeb111f0894f706af6c82187f3192ba3b327)

vicente-romero-oracle commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/LambdaToMethod.java:
```diff
@@ -321,7 +343,11 @@ public void visitClassDef(JCClassDecl tree) {
     public void visitLambda(JCLambda tree) {
-        LambdaTranslationContext localContext = (LambdaTranslationContext)context;
+        LambdaTranslationContext localContext = new LambdaTranslationContext(tree);
+        if (dumpLambdaToMethodStats) {
```
probably in a follow-up fix, but I don't see too much value on keeping / growing this stats code. This seems to belong in a regression test that probably uses a plugin or similar to get some stats while compiling lambdas / mrefs

mcimadamore replied:
Actually... in an attempt to remove the translation context classes (and just keep the lambda translation context class - after all method references don't need much there), I stumbled upon a bad test failure in the lambda deduplication test. This test depends on the compiler generating the stat diagnostics for every encountered lambda... maybe we can drop that dependency and the stats, but probably better to do it elsewhere.

vicente-romero-oracle commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/LambdaToMethod.java:
```diff
@@ -916,809 +952,335 @@ private JCExpression makeIndyCall(DiagnosticPosition pos, Type site, Name bsmNam
+                    propagateAnnos = false;
+                    break;
+                case LOCAL_VAR:
+                    ret = new VarSymbol(sym.flags() & FINAL, sym.name, sym.type, translatedSym);
```
shouldn't this symbol be synthetic too? same for the param below

mcimadamore replied:
I suppose local variables and parameters come from "real" symbols in the source code. We could probably do either way, but for now I preferred not to touch this stuff

mcimadamore commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/CaptureScanner.java:
```diff
@@ -0,0 +1,101 @@
+     * The list of owner's variables accessed from within the local class,
+     * without any duplicates.
+     */
+    private final ListBuffer<VarSymbol> fvs = new ListBuffer<>();
```
note that we use `prepend` to add symbols (which is what the old code used). So, we *could* change to something else, but w/o prepending, we'd create even more ordering issues. Maybe investigate separately?

mcimadamore commented:
> After trying different approaches, I realized that the cleanest (by far) solution is to simply capture the lambda owner symbol during `Attr`, instead of trying to reconstruct this owner using various (wrong) heuristics. When doing this, I realized that there were several issues which predated my refactoring - that is, the logic in `TypeAnnotations` relies on the `BLOCK` flag to be set on static/instance initializer owner method symbols - but sometimes this flag was missing (this is now fixed). Also, the `EnclosingMethodAttribute` sometimes reported `<clinit>` being the method owner of a local class. This is deliberately against the JVMS, which state:
> 
> > In particular, method_index must be zero if the current class was immediately enclosed in source code by an instance initializer, static initializer, instance variable initializer, or class variable initializer. (The first two concern both local classes and anonymous classes, while the last two concern anonymous classes declared on the right hand side of a field assignment.)
> 
> We even had a test which enforced the _wrong_ behavior (!!).
> 
> Last, the static/instance initializer symbol used to have a `null` type. This is now fixed, as we can always rely on the `BLOCK` flag instead (which is, again, cleaner).
> 
> As a result, `CaptureSiteInfo` is now gone.

I've ended up rewriting a biggie chunk in `Attr::lambdaEnv` as that seemed out of sync with what the rest of the compiler did (and also found other instances where `EnclosingMethod` contained a method index for lambdas **not** declared inside constructors)

vicente-romero-oracle commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/LambdaToMethod.java:
```diff
@@ -794,16 +830,16 @@ private VarSymbol makeSyntheticVar(long flags, Name name, Type type, Symbol owne
     private MethodType typeToMethodType(Type mt) {
         Type type = types.erasure(mt);
```
not part of your patch but I think that `mt` should already be erased at this point, now that Lower is running before L2M

mcimadamore replied:
I don't think so - I see it used here:

```
MethodSymbol samSym = (MethodSymbol) types.findDescriptorSymbol(tree.target.tsym);
List<LoadableConstant> staticArgs = List.of(
                typeToMethodType(samSym.type),
                refSym.asHandle(),
                typeToMethodType(tree.getDescriptorType(types)));
```

I think neither `samSym.type` nor the result of `tree::getDescriptorType` (which is based on `tree.target`) are erased.

vicente-romero-oracle reviewed (approved):
looks good, great fix!

mcimadamore commented:
Found a way to simplify the code some more by dropping the `TranslationContext` superclass and keep the `DeduplicationTest` happy. This should be the last cleanup. @cushon things look stable on our end, maybe you want to give this PR another try?

cushon commented:
> @cushon things look stable on our end, maybe you want to give this PR another try?

I will verify with the latest changes, but I have preliminary results from testing before https://github.com/openjdk/jdk/pull/20349/commits/6a5501fa83aef2ee455e7c965a1ebc74d36b97ec. There were no unexpected regressions, and some minor expected compatibility impact:

* A number of tests are asserting on names of synthetic lambda methods that have changed (e.g. `assertThat(Throwables.getStackTraceAsString(...)).contains("foo$0");`)
* I found a single test that was reading in persisted serialized data which no longer deserialized, and had to be updated to match the new output

I also removed the workarounds I'd added for JDK-8336492 and confirmed that they are unnecessary after these changes.

mcimadamore commented:
> * A number of tests are asserting on names of synthetic lambda methods that have changed (e.g. `assertThat(Throwables.getStackTraceAsString(...)).contains("foo$0");`)
> 
>     * I found a single test that was reading in persisted serialized data which no longer deserialized, and had to be updated to match the new output

Thanks. I suppose we have a good starting point then. If this proves to be too much incompatibility we could also try to fixup the logic later so that the generated names are more similar to the ones we had. It would be interesting if you could share some details about the deserialization failure: a redundant `this$0` in the serialized stream should be handles (serialization allows to ignore and reorder fields). Was there a name incompatibility in the captures?

cushon commented:
> It would be interesting if you could share some details about the deserialization failure: a redundant `this$0` in the serialized stream should be handles (serialization allows to ignore and reorder fields). Was there a name incompatibility in the captures?

I investigated, it looks like the order of a captured local and parameter is now reversed in the signature:

```java
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;
import java.nio.file.Files;
import java.nio.file.Path;

public class X {
  public static void main(String[] args) throws Exception {
    final Path path = Path.of(args[1]);
    switch (args[0]) {
      case "write" -> write(path, "hello");
      case "read" -> read(path);
    }
  }

  static void write(Path path, String s) throws Exception {
    int x = 42;
    Runnable r =
        (Runnable & Serializable)
            () -> {
              System.err.println(s + x);
            };
    try (ObjectOutputStream oos = new ObjectOutputStream(Files.newOutputStream(path))) {
      oos.writeObject(r);
    }
  }

  static void read(Path path) throws Exception {
    Runnable r;
    try (ObjectInputStream ois = new ObjectInputStream(Files.newInputStream(path))) {
      r = (Runnable) ois.readObject();
    }
    r.run();
  }
}
```

```
$BASELINE/bin/javac --release 22 X.java && $BASELINE/bin/java X write o && $EXPERIMENT/bin/javac --release 22 X.java && $EXPERIMENT/bin/java X read o
...
Exception in thread "main" java.io.InvalidObjectException: ReflectiveOperationException during deserialization
        at java.base/java.lang.invoke.SerializedLambda.readResolve(SerializedLambda.java:280)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at java.base/java.io.ObjectStreamClass.invokeReadResolve(ObjectStreamClass.java:1198)
        at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2278)
        at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1747)
        at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:525)
        at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:483)
        at X.read(X.java:31)
        at X.main(X.java:12)
Caused by: java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:118)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at java.base/java.lang.invoke.SerializedLambda.readResolve(SerializedLambda.java:278)
        ... 9 more
Caused by: java.lang.IllegalArgumentException: Invalid lambda deserialization
        at X.$deserializeLambda$(X.java:7)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        ... 11 more
```

Comparing the output shows:

```diff
494,495c494,495
<   private static void lambda$write$e26d8c1c$1(java.lang.String, int);
<     descriptor: (Ljava/lang/String;I)V
---
>   private static void lambda$write$b96aa0a6$1(int, java.lang.String);
>     descriptor: (ILjava/lang/String;)V
```

vicente-romero-oracle reviewed (approved):
latest changes looks good too

mcimadamore commented:
> I investigated, it looks like the order of a captured local and parameter is now reversed in the signature:

I see - since the serialized form of a lambda also includes a signature string, we fail to deserialize if the signature string cannot be parsed to a `MethodType` that is consistent with the type of the generated lambda method. Also, since the order of captures is different, the serialized lambda name is also different.

The order of captures has changed because `CaptureScanner`. Now, if I compile this program using Java 22:

```java
class Test {
   void m(String s, int i) {
       new Object() {
           void g() { System.out.println(s + i); }
       };
   }

   Runnable r(String s, int i) {
       return () -> System.out.println(s + i);
   }
}
```
I see that the order of capture is the same in both the anon class constructor and the lambda:

```java
  private static void lambda$r$0(java.lang.String, int);
...
  Test$1(Test, java.lang.String, int);
```

But with this patch they are not:

```java
  private static void lambda$r$0(int, java.lang.String);
...
  Test$1(Test, java.lang.String, int);
```

The peculiarity (which I missed) is that `Lower` gets the set of freevars, and then always _prepend_ their definition to the class/constructor. `LambdaToMethod`, OTOH, _appends_ them. This is what's causing the mismatched order. I'll fix this, as I think the old order looked better than the new one (as each captured variable appear in "order of capture").

mcimadamore commented:
I've added a couple of changes:
* first, I made sure that the order of captured local is the same as before (and, importantly, that the order is consistent across local classes and lambdas)
* second, I reverted the changes to drop `this$0` in serializable local classes. Thinking more about this, this seemed to me mostly an orthogonal improvement, unrelated to the bulk of this fix. If we want to pursue this we should probably (a) do that in a separate issue and (b) do that for both local and member classes (after checking the JLS, there's nothing there saying that member classes should have a _field_, or that the name of such field is specified in any way).

This means that this PR leaves us with two minor incompatibilities:
* the name of non-serializable lambdas might change (for now, I'd like to keep it this way, as I think new names are better, and relying on unspecified generated names is discouraged anyways - we can revisit if absolutely necessary)
* due to the fixes in this PR, javac now correctly omits the method index in some of the generated `EnclosingMethod` attributes, when the lambda is defined inside a static or instance initializer.

vicente-romero-oracle commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/CaptureScanner.java:
```diff
@@ -0,0 +1,99 @@
+    private final Set<Symbol.VarSymbol> seenVars = new HashSet<>();
+
+    /**
+     * The list of owner's variables accessed from within the local class,
```
I think that now that the field is a set we probably should update this comment, it is not a list anymore and it can't have duplicates

vicente-romero-oracle reviewed (approved):
latest changes look good

cushon commented:
I did another round of testing with the latest version of the PR, everything still looks good to me.

* I am no longer seeing any serialization compatibility impact.
* I'm still seeing some impact from changes to lambda names. The impact is minor, and all of the occurrences I saw were in tests. I don't have any concerns about proceeding with that part.

mcimadamore commented:
> I did another round of testing with the latest version of the PR, everything still looks good to me.
> 
>     * I am no longer seeing any serialization compatibility impact.
> 
>     * I'm still seeing some impact from changes to lambda names. The impact is minor, and all of the occurrences I saw were in tests. I don't have any concerns about proceeding with that part.

Thanks for confirming. I think this is in good shape, and we can now proceed.

ExE-Boss commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/CaptureScanner.java:
```diff
@@ -54,10 +54,9 @@ public class CaptureScanner extends TreeScanner {
+     * The set of captured local variables accessed from within the tree under analysis.
      */
-    private final ListBuffer<VarSymbol> fvs = new ListBuffer<>();
+    private final Set<VarSymbol> fvs = new LinkedHashSet<>();
```
```suggestion
    private final SequencedSet<VarSymbol> fvs = new LinkedHashSet<>();
```

ExE-Boss commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/CaptureScanner.java:
```diff
@@ -96,6 +93,6 @@ public void visitVarDef(JCTree.JCVariableDecl tree) {
     List<Symbol.VarSymbol> analyzeCaptures() {
         scan(tree);
-        return fvs.toList();
+        return List.from(fvs);
```
```suggestion
        return List.from(fvs.reversed());
```

ExE-Boss commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Lower.java:
```diff
@@ -314,7 +314,7 @@ List<VarSymbol> freevars(ClassSymbol c)  {
         }
         FreeVarCollector collector = new FreeVarCollector(classDef(c));
-        fvs = collector.analyzeCaptures();
+        fvs = collector.analyzeCaptures().reverse();
```
```suggestion
        fvs = collector.analyzeCaptures();
```

ExE-Boss commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/CaptureScanner.java:
```diff
@@ -0,0 +1,98 @@
+
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.Set;
```
```suggestion
import java.util.Set;
import java.util.SequencedSet;
```

ExE-Boss reviewed (commented):
Using `SequencedSet` would allow to skip the call to `com.sun.tools.javac.util.List​::reverse()` on the result of `CaptureScanner​::analyzeCaptures()`.

mcimadamore commented:
> Using `SequencedSet` would allow to skip the call to `com.sun.tools.javac.util.List​::reverse()` on the result of `CaptureScanner​::analyzeCaptures()`.

Not sure about that. Note that `Lower` wants the order of capture to be reversed compared to the "logical" order, but `LambdaToMethod` does not. To me, it is cleaner to put the reverse call close to the code that has "special" requirements, which in this case happens to be `Lower` (which is why the code ended up with that shape).

ExE-Boss commented:
> > Using `SequencedSet` would allow to skip the call to `com.sun.tools.javac.util.List​::reverse()` on the result of `CaptureScanner​::analyzeCaptures()`.
> 
> Not sure about that. Note that `Lower` wants the order of capture to be reversed compared to the "logical" order, but `LambdaToMethod` does not. To me, it is cleaner to put the reverse call close to the code that has "special" requirements, which in this case happens to be `Lower` (which is why the code ended up with that shape).

Well, at least typing the field as a `SequencedSet` would signal that the iteration order is important.

mcimadamore commented:
/integrate

mcimadamore commented:
/integrate

vicente-romero-oracle reviewed (approved):
looks good

mcimadamore commented:
/integrate
